server/server.go

```diff
@@ -893,9 +940,154 @@ func (s *Server) v1QueryGet(w http.ResponseWriter, r *http.Request) {
 	writer.JSON(w, 200, results, pretty)
 }
 
+func constructDiagnosticsInput(r *http.Request) (map[string]interface{}, error) {
+	var body interface{}
+	if r.Body != nil {
+		buf, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return nil, err
+		}
+		r.Body = ioutil.NopCloser(bytes.NewReader(buf))
+
+		decoder := util.NewJSONDecoder(bytes.NewReader(buf))
+		if err := decoder.Decode(&body); err != nil {
+			body = nil
+		}
+	}
+
+	input := map[string]interface{}{
+		"method": r.Method,
+		"path":   r.URL.Path,
+		"body":   body,
+	}
+
+	params := map[string]interface{}{}
+	for param, value := range r.URL.Query() {
+		var ifaces []interface{}
+		for _, v := range value {
+			ifaces = append(ifaces, v)
+		}
+		params[param] = ifaces
+	}
+	input["params"] = params
+
+	header := map[string]interface{}{}
+	for key, value := range r.Header {
+		var ifaces []interface{}
+		for _, v := range value {
+			ifaces = append(ifaces, v)
+		}
+		header[key] = ifaces
+	}
+	input["header"] = header
+
+	return input, nil
+}
+
+func (s *Server) evalDiagnosticPolicy(r *http.Request) diagnostics.Settings {
```

Comment:
Do you know how much overhead this adds? I'm curious. If it's not negligible for a trivial policy then leave it as is.